### DISCUSSION TITLE
Can we exclude the some accessible namespaces in kiali CR with some labelSelector?

### DISCUSSION THREAD

**Discussion by amantayal5** [kiali/kiali#5470] [Q&A]
We are using Kiali server model to deploy kiali and we want to exclude some namespaces in the cluster which have a certain label.
Can we add this capability? Or if it is already present, can you redirect me to its documentation as I couldn't find it?

**Comment by jmazzitelli** [comment]
_converted to Q&A Discussion_

**Comment by jmazzitelli** [comment]
See: https://kiali.io/docs/configuration/namespace-management/#namespace-selectors

**Reply by jmazzitelli** [reply]
Note that because you are using the [non-recommended server helm chart](https://kiali.io/docs/installation/installation-guide/install-with-helm/#introduction), anything you read in the docs about "accessible_namespaces" not being set to "**" and the features that arise from that (e.g. labels getting created on namespaces for you) are not applicable. The operator is the one that does the special processing for this `accessible_namespaces` setting.

That said, the Kiali server will honor the `api.namespaces.label_selector` so that [portion of the docs I linked to above](https://kiali.io/docs/configuration/namespace-management/#namespace-selectors) is what you want.

**Reply by amantayal5** [reply]
I understand that api.namespaces.label_selector makes sure kiali CR to have access to all the namespaces with particular label but we were thinking can we have this in api.namespaces.exclude.label_selector?

Our major use case is to restrict application developers to have access to the platform namespaces.

**Reply by jmazzitelli** [reply]
> can we have this in api.namespaces.exclude.label_selector

Two things. First, that `api.namespaces.exclude` setting is not security related, it is merely used as a filter but doesn't truly provide security. See the [docs](https://kiali.io/docs/configuration/namespace-management/#excluded-namespaces) where it says, `This setting has no effect on namespace accessibility. It is only a filter, not security-related.` If your concern is to `restrict application developers to have access to the platform namespaces` then your concern is security related and thus it won't do what you are expecting. So even if Kiali supported labels, as opposed to regex of namespace names, it wouldn't be what you want.

Second, what you want is already supported. You can do this in one of two ways.

The first option is you can set `spec.deployment.accessible_namespaces` in the Kiali CR and let the operator create roles that restrict the namespaces that Kiali can access. Doing this means all users of Kiali can be restricted to access only those namespaces you provide in that accessible_namespaces setting. That setting supports regex, so you can define things like "allow access to all namespaces except namespace kube-system" (indeed, that is basically what [the default is already](https://github.com/kiali/kiali-operator/blob/master/crd-docs/cr/kiali.io_v1alpha1_kiali.yaml#L59) - Kiali is restricted from seeing all internal namespaces that start with `kube.` and things like that).

 The second option is use RBAC. After all, this is what you want - you want certain users to be restricted to see certain resources. This is the classic RBAC use case. So, in the Kiali CR, set a specific `auth.strategy` that is appropriate for your platform (in other words, do not use `anonymous`). Most people employing RBAC are using OpenID - so set `auth.strategy` to `openid` and configure Kiali to work with your OpenID provider. At this point, it is just a matter of setting up your application developer's kubernetes user roles and permissions just like you would for anything else. For docs on Kiali's RBAC support and its different auth strategies, see https://kiali.io/docs/configuration/authentication/

**Reply by amantayal5** [reply]
@jmazzitelli I understand `api.namespace.exclude` setting is not security related and its just used as filter.
We are using kiali in **read-only** mode and what I was asking is having **label-selector** field in `api.namespace.exclude` section. It helps filter the namespaces in much easier way.
We already have a functionality in Kiali CR to filter accessible namespaces using label selector in `api.namespace.label-selector`, I was wondering having a similar thing to exclude will also be very helpful!

**Reply by jmazzitelli** [reply]
enhancement request submitted: https://github.com/kiali/kiali/issues/5516
